Add template generation to DaemonSet upgrade proposal

janetkuo opened the pull request:
Add an additional field `.templateGeneration` to DaemonSet, to keep track of DaemonSet template changes. We need it because we need to compare DaemonSet template with DaemonSet pods to know if the pods need to be updated, and:

1. Pod template hash isn't 100% reliable - it may change on PodTemplate API changes (e.g. adding new fields), and users will encounter unexpected DeamonSet rolling update after upgrading Kubernetes to a different version. 
2. It's not easy to infer pod template from a pod's spec, since many default values are applied to the pod, and the admission controller may update the pod as well. 

@erictune @lavalamp @kargakis @lukaszo @smarterclayton @kubernetes/sig-apps-api-reviews @kow3ns @foxish @enisoc @pwittrock 

--- 
Copied from my comment in https://github.com/kubernetes/kubernetes/pull/41116#issuecomment-279817775:

1. In DaemonSet update strategy, PodTemplateGeneration++ on DS.spec.template updates
2. When DS creates pods, label them with current PodTemplateGeneration ("pod-template-generation=n")
    - Use labels since it's easy for the controller and the users to filter pods that match current podtemplate generation.
3. DS compares pods' "pod-template-generation" label against its own PodTemplateGeneration to decide to kill (update) this pod or not.

One drawback is that it's not extensible (customized controllers that extend DaemonSet controller will have to modify API server code as well) - we decided not to worry about this for now. 

Another drawback is that in the rollback case (in the future), we may restart more pods than we want (since we can only roll forward PodTemplateGeneration), and pods gen 1 may actually be the same as DS gen 3. 

We can solve the rollback case with DaemonSet history + pod template hash. We use history objects (maybe stand-alone PodTemplate objects?) to store pod-template-hash and templates, and we compare pods' pod-template-hash labels with history and compare history template with DaemonSet template. 

**Alternatives:** 
1. I thought about using `json.Marshal` before hashing the object so that new fields with `omitempty` wouldn't change the hash value. However, if a new field was added to PodTemplate with some default value, the hash will still change. 
2. White-listing fields we care about when hashing pod template... this is not extensible. 
3. Implement a helper function that gives you podtemplate given a pod spec. This is hard to implement, as many default values will be added to a pod, and admission controller will also update the pod.

smarterclayton commented:
If we have history, do we still need the hash?  I would expect us to deep equal pod / history / workflow object

kargakis commented:
> If we have history, do we still need the hash? I would expect us to deep equal pod / history / workflow object

No I don't think we need it. We definitely need to stop hashing the podtemplate but I was thinking of hashing the deployment name + new generation to keep the replica set names as is. May make sense for PodTemplates too but I wouldn't expect those to be as much involved so their names may not matter that much.

kargakis commented on contributors/design-proposals/daemonset-update.md:
```diff
@@ -129,18 +127,35 @@ type DaemonSetStatus struct {
 	NumberUnavailable int32 `json:"numberUnavailable"`
 }
+
+type DaemonSet struct {
```
Should be in DaemonSetSpec

janetkuo replied:
The reason that I put it under `DaemonSet` instead of `DaemonSetSpec` is that `TemplateGeneration` is read-only and generated by system - wouldn't it be a bit odd that such field exists in the spec, where users will specify expected state there?

kargakis replied:
Hm, there is no precedent for this - @kubernetes/api-reviewers can you weigh in?

janetkuo replied:
cc @bgrant0607 @pwittrock @erictune

kow3ns replied:
Could we consider putting the field under the DaemonSetStatus? DeamonSetStatus is generally mutable by the controller, and used to communicate information about the Status of the objects corresponding control loop. My thought is that the current templateGeneration that the DaemonSet controller is attempting to conform its Pods to is part the status of the controller.

janetkuo replied:
> Could we consider putting the field under the DaemonSetStatus?

Not sure how to do that in API server. [Update strategy](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/extensions/daemonset/strategy.go#L58-L80) forbids updating status, and [status update strategy](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/extensions/daemonset/strategy.go#L141-L145) ignores changes not in status. Also, we update DaemonSet status in DaemonSet controller sync loop, not in API server. 

Maybe we could update templateGeneration in status when we get notification from informer that DaemonSet template is updated, assuming informer is reliable and never misses any event?

kargakis commented on contributors/design-proposals/daemonset-update.md:
```diff
@@ -129,18 +127,35 @@ type DaemonSetStatus struct {
+	// DaemonSetTemplateGenerationKey is the key of the labels that is added
+	// to daemon set pods to distinguish between old and new pod templates
+	// during DaemonSet template update.
+	DaemonSetTemplateGenerationKey string = "pod-template-generation"
```
If there is no strong argument against it, let's make this namespaced.

lukaszo replied:
We still don't know why in similar case in Deployments namespace was removed. @bgrant0607 can you comment on it?

janetkuo replied:
Namespace was removed from "pod-template-hash" label in #20765. The main reason is that `pod-template-hash` is more suitable as a label name than `deployment.kubernetes.io/podTemplateHash`. `deployment.kubernetes.io/podTemplateHash` is more like an annotation, not label.

kargakis replied:
Pretty sure we are using namespaced labels all over the place for core components.

janetkuo replied:
See https://github.com/kubernetes/kubernetes/pull/20765#r52097671, 

> Discussed with @bgrant0607 about this, and we renamed it to be without namespace because it's too long whenever we display resources with their labels, and putting namespace is more suitable for annotations because annotations targets machines, not humans.

kargakis replied:
Why are namespaced labels targetting machines only? I thought that the incentive behind namespacing {labels, annotations} is to cap on those label keys and make sure users won't fiddle with them.

janetkuo replied:
See Brian's previous comment:
> The desire for lack of a namespace prefix was one reason I had originally insisted on the key being configurable -- so that the user could change the key in order to avoid key-name collisions with their own labels.
However, in the short term it's easier if we don't support that field. If users request it, we can add it back later.

We should instead make the label key configurable to avoid collisions. 

Because we want the label key to be non-namespaced, we originally designed Deployment "pod template hash" label key to be configurable for users to avoid key name collisions. We haven't, however, implement that in Deployment, since we haven't see no user request for that yet.

kargakis replied:
After revising the Deployment proposal (https://github.com/kubernetes/community/pull/384) I realized that we don't need the extra label in the PodTemplate for Deployments since now ReplicaSets set owner references in Pods. It will be needed for DaemonSets though since there is no other controller below it to set owner references thus separate between two different versions of Pods.

kargakis commented on contributors/design-proposals/daemonset-update.md:
```diff
@@ -158,9 +173,10 @@ For each pending DaemonSet updates, it will:
-       `.spec.template.spec` to see if DaemonSet spec has changed.
-     - If DaemonSet spec has changed, compare `MaxUnavailable` with DaemonSet
+     - Find all daemon pods that belong to this DaemonSet using label selectors.
+       - Pods with "pod-template-generation" value equal to the DaemonSet's
```
What about pods that don't know about this label ie. all existing pods?

janetkuo replied:
Did you mean the pre-1.6 pods? Since we don't have a way to compare pod spec against DaemonSet template, we can only kill them all (they'll get updated with right labels) or ignore them (no update). Killing them is probably not safe. Ignoring them means users will need to manually kill pre-1.6 pods to enable DaemonSet rolling update (if the DaemonSet was created before 1.6).

lukaszo replied:
We can try to compare pod template with ds.Spec.Template. In many cases it will just work. In other cases we may generate event or at least log it.

kow3ns replied:
Could we use a nil template generation in order to indicate we expect the label to not be set? In this case the DaemonSet does not attempt the comparison until the first time the upgrade is performed.

janetkuo replied:
> We can try to compare pod template with ds.Spec.Template. 

But we don't have pod template to compare with DaemonSet template in 1.6. It only exists in history. 

> Could we use a nil template generation in order to indicate we expect the label to not be set? In this case the DaemonSet does not attempt the comparison until the first time the upgrade is performed.

In current proposal, template generation is read-only and only modified by system. Users have no control over it so can't specify their intent through it.

kargakis replied:
I think kicking off a rollout just to relabel the pods once users switch to use the Rolling strategy is fine as long as we have the behavior documented.

janetkuo replied:
> I think kicking off a rollout just to relabel the pods once users switch to use the Rolling strategy is fine as long as we have the behavior documented.

I'm +1 for kicking off a rollout (killing them actively), since they won't be surprised by default (they have to modify update strategy to trigger this).

kargakis commented on contributors/design-proposals/daemonset-update.md:
```diff
@@ -270,22 +292,40 @@ Another way to implement DaemonSet history is through creating `PodTemplates` as
+   `.spec.revisionHistoryLimit` latest PodTemplates (remove the rest).
+  - Find the PodTemplate whose "pod-template-generation" label value equals to
+    DaemonSet's `.templateGeneration`. If not found, find the PodTemplate whose
+    "pod-template-hash" label value equals to the hash value of DaemonSet's
```
Why do we need pod-template-hash? I think of this proposal as a better alternative to hashing and since both new labels need to be added we are ok with one of them (specifically the one about generation).

kow3ns replied:
Agree, I think that hash should go if we use this method.

redbaron commented:
> Another drawback is that in the rollback case (in the future), we may restart more pods than we want (since we can only roll forward PodTemplateGeneration), and pods gen 1 may actually be the same as DS gen 3.

This is very unfortunate, this means that rollback is actually a force-restart of all pods even if rollback is triggered by canary.

Can't it be solved if `PodTemplateGeneration` is not a  number, but an always array of UUIDs, one per generation, which acts as a stack?  New generation is expressed as stack push and rollback is stack pop?

kargakis commented:
> Another drawback is that in the rollback case (in the future), we may restart more pods than we want (since we can only roll forward PodTemplateGeneration), and pods gen 1 may actually be the same as DS gen 3.

We kill pods so there is no restarting involved, right? If we want to rollback from 2 to 1, we spin up any missing pods and relabel the remaining from 1 to 3.

janetkuo commented:
> We kill pods so there is no restarting involved, right? If we want to rollback from 2 to 1, we spin up any missing pods and relabel the remaining from 1 to 3.

By "restarting" I meant "killing and recreating pods with the same template". It depends on how we implement rollback. We bump `TemplateGeneration` in API server (not in controller sync loop), how would DaemonSet controller know that gen 3 = gen 1? We'll need to store such information somewhere - preferably not in memory. 

> Can't it be solved if PodTemplateGeneration is not a number, but an always array of UUIDs, one per generation, which acts as a stack? New generation is expressed as stack push and rollback is stack pop?

Stack push and pop is an interesting idea. However, IIUC, it only handles rolling back to last version? What if we want to rollback from v10 to v1?

lukaszo commented:
2017-02-15 19:27 GMT+01:00 Janet Kuo <notifications@github.com>:

> We kill pods so there is no restarting involved, right? If we want to
> rollback from 2 to 1, we spin up any missing pods and relabel the remaining
> from 1 to 3.
>
> By "restarting" I meant "killing and recreating pods with the same
> template". It depends on how we implement rollback. We bump
> TemplateGeneration in API server (not in controller sync loop), how would
> DaemonSet controller know that gen 3 = gen 1? We'll need to store such
> information somewhere - preferably not in memory.
>
I think it will look like this:

- user wants rollback to revision 1
- controller creates new revision let's say 3
- controller finds all pods with revision 1 and changes label to 3
- controller kills all remaining pods



> Can't it be solved if PodTemplateGeneration is not a number, but an always
> array of UUIDs, one per generation, which acts as a stack? New generation
> is expressed as stack push and rollback is stack pop?
>
> Stack push and pop is an interesting idea. However, IIUC, it only handles
> rolling back to last version? What if we want to rollback from v10 to v1?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/366#issuecomment-280095540>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAF8N1DWAgzgsnCOfo4ZBHDhX2FgrvXhks5rc0N7gaJpZM4MBKan>
> .
>



-- 
Łukasz Oleś

redbaron commented:
> @redbaron Stack push and pop is an interesting idea. However, IIUC, it only handles rolling back to last version? What if we want to rollback from v10 to v1?


High level loop is simple.  One every event do:
 0. update desired generations stack
 1. every pod from generation which is found in the stack, but not at the top of it  is added to "podsToUpdate" list
 2. every pod from generation which is not found in the stack is added to "podsToRollback" list
 3. every pod from generation which is at the top of the stack is added to "podsToKeep" list

these lists are then handled by (possibly pluggable) `UpdateStrategy`, `RollbackStrategy` and `KeepStrategy` "mini-controllers".  They issue commands to API server to create or kill pods.

so it doesn't matter how much back you need to rollback, you can converge cluster state avoiding unnecessary restarts.

kargakis commented:
> @kargakis By "restarting" I meant "killing and recreating pods with the same template". It depends on how we implement rollback. We bump TemplateGeneration in API server (not in controller sync loop), how would DaemonSet controller know that gen 3 = gen 1? We'll need to store such information somewhere - preferably not in memory.

Compare the respective PodTemplates? I don't know how are we going to implement rollback without history if this is assumed.

kargakis commented:
Another thing worth noting is that templateGeneration does not equal revision meaning the controller may miss some generations. So I would expect that admins would rollback based on the revision of the DaemonSet (the generations that were materialized into templates) and not the TemplateGeneration.

janetkuo commented:
So, 
- Remove pod template hash. // I kept it only because it was in the original proposal and it's more efficient to list all pods with a specific template at once (via selector), but it's not required.
- _(future)_ DeepEqual history templates with DaemonSet template during rollback/rolling update.

Questions on using revision of the DaemonSet instead of templateGeneration:
- Where is DaemonSet revision stored? In its history resources (e.g. each history will have a revision number, maybe a label or annotation)?
- If we'll depend on revision of the DaemonSet instead of templateGeneration, this means templateGeneration won't be a reliable source for template vs. revision mapping. Then how do we compare a pod with DaemonSet template in 1.6, without history? Do we use pod-template-generation labels in pods first, and then switch to revision labels in the future?

kargakis commented:
> Where is DaemonSet revision stored? In its history resources (e.g. each history will have a revision number, maybe a label or annotation)?

Each PodTemplate stores both the revision of the DaemonSet (as an annotation similar to how it's stored in ReplicaSets owned by Deployments) and the generation it represents.

> If we'll depend on revision of the DaemonSet instead of templateGeneration, this means templateGeneration won't be a reliable source for template vs. revision mapping. Then how do we compare a pod with DaemonSet template in 1.6, without history? Do we use pod-template-generation labels in pods first, and then switch to revision labels in the future?

TemplateGeneration is reliable, revision is not because it is set by the controller but we cannot use generation for rollbacks because some may be missed. Since we have no history in 1.6 there is no revision so we stay with generation for comparing pods with the ds template. When we get history, we will be able to compare the ds template directly with the rest of the pod templates rather than merely looking at labels.

janetkuo commented:
@kargakis sounds good. Will update the proposal accordingly.

janetkuo commented:
To summarize, remaining discussions:

1. Where should we place `.templateGeneration`? (See https://github.com/kubernetes/community/pull/366#discussion_r101354897)
   - `DaemonSet` (current proposal)
   - `DaemonSetSpec`
   - `DaemonSetStatus` 
2. Should we namespace the pod template generation label? (See https://github.com/kubernetes/community/pull/366#discussion_r101208893)
   - No (current proposal) 
   - Yes
3. How to handle pre-1.6 DaemonSet pods? They don't have the pod template generation label, and we can't compare their spec with DaemonSet template. (See https://github.com/kubernetes/community/pull/366#discussion_r101208966)
   - Just ignore them. Users will need to manually kill those pods to rolling update a DaemonSet. Once those pre-1.6 DaemonSet pods are recreated or updated users can rolling update the DaemonSet normally. 
   - Actively kill those pods (so that we'll recreate them with template generation labels). Need to document this behavior that switching to rollingUpdate strategy will trigger rolling update of old DaemonSets (even when templates aren't changed).

erictune commented:
About where to put a field that is set by the API server at creation/update time.  For comparison, for selector Generation in Jobs, it puts the generated value into the `.spec`, and blocks users from setting/changing it in validation.

erictune commented:
Regarding extensibility, janetkuo wrote: 

 > One drawback is that it's not extensible (customized controllers that extend DaemonSet controller will have to modify API server code as well) - we decided not to worry about this for now.

I'd ideally like to follow a pattern here that extension controllers can follow.  That is a complicated problem.  I'm not sure either our 1.6 solution or the 1.7 (history) solution works for extensions.   I've been thinking it and filed an issue about it: https://github.com/kubernetes/kubernetes/issues/41677

kargakis commented:
There are other fields in the spec users are not supposed to update directly such as rollback for Deployments and in the future Brian and I expect that canaries will use the same pattern.

kargakis commented:
> To summarize, remaining discussions:
> 
>     Where should we place .templateGeneration? (See #366 (comment))
>         DaemonSet (current proposal)
>         DaemonSetSpec
>         DaemonSetStatus
>     Should we namespace the pod template generation label? (See #366 (comment))
>         No (current proposal)
>         Yes
>     How to handle pre-1.6 DaemonSet pods? They don't have the pod template generation label, and we can't compare their spec with DaemonSet template. (See #366 (comment))
>         Actively kill those pods (so that we'll recreate them with template generation labels). Need to document this behavior -- switching to rollingUpdate strategy will trigger rolling update of pre-1.6 DaemonSets (even when templates aren't changed).
>         Just ignore them. Users will need to manually kill those pods to rolling update a DaemonSet. Once those pre-1.6 DaemonSet pods are recreated or updated users can rolling update the DaemonSet normally.

1. Spec I believe is the best place
2. I would prefer namespaced but I don't feel strong
3. Kick-off a new rollout - the controller needs to know versions of pods once it starts managing upgrades. We need to make sure the behavior is documented.

janetkuo commented:
Updated proposal:

1. Put `.templateGeneration` field in `.spec`
2. I prefer not namespacing pod-template-generation label, so this part is unchanged 
3. Added a note that pre-1.6 pods will be killed (rolling update will be triggered) after upgrading k8s to 1.6 and setting the updateStrategy to RollingUpdate

Also note that `templateGeneration` will be increased on template changes, regardless of updateStrategy. That is to say, if users update their DaemonSet templates during "OnDelete" updateStrategy and then switch to "RollingUpdate" strategy, the rollout will also be triggered no matter if the template is the same as current DaemonSet pods or not.

erictune commented:
LGTM.

Try to make the adoption case work well (create a new DaemonSet which has RollingUpdate strategy which adopts existing pods which have a templateVersion, and it does not restart them because they are as expected.

erictune commented:
/approve

janetkuo commented:
Supporting pods adoption is a good reason to keep pod-template-hash. With merely templateGeneration, DaemonSets won't be able to adopt existing pods, unless we allow user-specified templateGeneration. If we use templateGeneration + template hash, DaemonSet controller can still find those pods they're supposed to adopt. (Note that I'm not talking about adopting pre-1.6 pods.) WDYT? @kargakis @smarterclayton @lukaszo

kargakis commented:
> That is to say, if users update their DaemonSet templates during "OnDelete" updateStrategy and then switch to "RollingUpdate" strategy, the rollout will also be triggered no matter if the template is the same as current DaemonSet pods or not.

This shouldn't be the case once we have history because now we trigger rollouts based on different templateGenerations, with history we will trigger rollouts based on differences between the PodTemplates.

kargakis commented:
> Supporting pods adoption is a good reason to keep pod-template-hash. With merely templateGeneration, DaemonSets won't be able to adopt existing pods, unless we allow user-specified templateGeneration. If we use templateGeneration + template hash, DaemonSet controller can still find those pods they're supposed to adopt. (Note that I'm not talking about adopting pre-1.6 pods.) WDYT? @kargakis @smarterclayton @lukaszo

A DaemonSet should be able to adopt any pod that matches its selector (we don't add pod-template-generation in the DS selector). If an adopted pod has no pod-template-generation label or a pod-template-generation value different from the current one the DS has, then it's going to be killed.

kargakis commented:
> If an adopted pod has no pod-template-generation label or a pod-template-generation value different from the current one the DS has, then it's going to be killed.

For 1.6. Assuming in 1.7 we implement history, this should change to "look at the value of pod's pod-template-generation label, fetch respective PodTemplate, compare with ds.spec.template, if equal, relabel pod to the latest generation, create new template with both revisions, remove old duplicate template"

janetkuo commented:
Imagine the pod adoption like this (in 1.6):

1. We have a DaemonSet A, its update strategy is "RollingUpdate"
2. At some point its templateGeneration=3, and all its pods are labeled "pod-template-generation=3"
3. We kill A, but leave its pods orphaned. 
4. We create a DaemonSet B, which is exactly the same as A. But now B's templateGeneration=1.
5. DaemonSet controller kills all pods that should've been adopted by B, because their labels are "pod-template-generation=3" and not equal to B's templateGeneration. Oops. 

With pod template hash, we can as least calculate B's template hash and compare that with those pods' "pod-template-hash" label.

kargakis commented:
I guess we can use pod-template-hash for allowing renaming but it will be
superseded once we have history and we should stop using it so I don't feel
strong about it.

On Wed, Feb 22, 2017 at 8:05 AM, Janet Kuo <notifications@github.com> wrote:

> Imagine the pod adoption like this (in 1.6):
>
>    1. We have a DaemonSet A, its update strategy is "RollingUpdate"
>    2. At some point its templateGeneration=3, and all its pods are
>    labeled "pod-template-generation=3"
>    3. We kill A, but leave its pods orphaned.
>    4. We create a DaemonSet B, which is exactly the same as A. But now
>    B's templateGeneration=1.
>    5. DaemonSet controller kills all pods that should've been adopted by
>    B, because their labels are "pod-template-generation=3" and not equal to
>    B's templateGeneration. Oops.
>
> With pod template hash, we can as least calculate B's template hash and
> compare that with those pods' "pod-template-hash" label.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/366#issuecomment-281587334>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADuFf4ZW72Krr_kD6rN8oOv46A-KHUP_ks5re946gaJpZM4MBKan>
> .
>

janetkuo commented:
Actually, we don't need `templateGeneration` once we have history. If `templateGeneration` will be reset every time a DaemonSet is deleted and recreated, it's not useful and we should instead look at existing history to reason about revisions. (Maybe we can look at current history and bump `templateGeneration` in update strategy, but it's messy.)

This then brings another question: how do we make sure we can always map pods to history? We can't  use history revision, because it may change when we do a rollback, and we can't guarantee that we'll always update pods' and history revisions together (updating history revision and pod revision are two separate steps and won't always succeed). 

Then we should use something more stable. "pod-template-hash" is an option, since we won't change the "pod-template-hash" of the history, and it's rare we'll upgrade the cluster during DaemonSet rolling update.

janetkuo commented:
We may also end up ignoring updating old pods if templateGeneration isn't reliable.

kargakis commented:
Hm, I have been thinking more in terms of Deployments... Hashing the DS PodTemplate and using pod-template-hash in Pods shouldn't be a big problem for DaemonSets because new PodTemplates won't trigger another process to spin up new Pods (such as the replica set controller is doing every time a new RS is created). The worst thing that can happen after a cluster upgrade is end up with a duplicate PodTemplate. With that in mind now I think we probably don't need TemplateGeneration for DS at all.

janetkuo commented:
Just discussed with @bgrant0607 and @erictune, we decided that we'll use `templateGeneration`, but allow users to set its value on creation time (and only on creation time) - this allows DaemonSet pods adoption to work as expected. 

None of the solutions is optimal without history. Things we thought about:
1. Use strategic merge diff to compare DaemonSet template against pod spec. However, we won't be able to detect fields deletion on DaemonSet templates. 
2. Use pod template hash. Then when users upgrade from 1.6 to any future release, DaemonSet upgrade may be triggered as well (due to hash value change, and we don't have history in 1.6). 
3. DaemonSet controller figure out the right templateGeneration from observing current pod template-generation label and add its value to 1. This is likely trigger unnecessary upgrades.

janetkuo commented:
This is LGTM'ed, merging

kargakis commented:
> Then when users upgrade from 1.6 to any future release, DaemonSet upgrade may be triggered as well (due to hash value change, and we don't have history in 1.6).

This shouldn't happen, see [why above](https://github.com/kubernetes/community/pull/366#issuecomment-281791829). The duplicate PodTemplate that got created from the cluster upgrade still equals the PodTemplate the running pods were created from so there should be no rollout of new pods.

kargakis commented:
> This shouldn't happen, see why above. The duplicate PodTemplate that got created from the cluster upgrade still equals the PodTemplate the running pods were created from so there should be no rollout of new pods.

Ah, nevermind. This can happen due to not having history. Settable templateGeneration on creation should work.
